compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -470,41 +533,75 @@ buildLfPackageGraph targetLfVersion pkgs stablePkgs dependencyPkgs = (depGraph,
         , LF.Qualified { LF.qualPackage } <- Set.toList quals
         ]
 
-    -- order the packages in topological order
     (depGraph, vertexToNode, _keyToVertex) =
         graphFromEdges
-            [ (PackageNode src decodedUnitId decodedDalfPkg, LF.dalfPackageId decodedDalfPkg, pkgRefs)
-            | DecodedDalf{decodedUnitId, decodedDalfPkg} <- pkgs
-            , let pkg = LF.extPackagePkg (LF.dalfPackagePkg decodedDalfPkg)
-            , let pkgRefs = [ pid | LF.PRImport pid <- allPackageRefs pkg ]
-            , let src = generateSrcPkgFromLf (config (LF.dalfPackageId decodedDalfPkg) decodedUnitId) pkg
+          [ (node, key, deps <> etc)
+          | v <- vertices depGraph0
+          , let (node, key, deps) = vertexToNode0 v
+          , let etc = case node of
+                  MkDataDependencyPackageNode {} -> depsWithoutDataDeps
+                  _ -> []
+          ]
+
+    depsWithoutDataDeps =
+      [ key
+      | v <- vertices depGraph0
+      , (node, key, _) <- [vertexToNode0 v]
+      , isDependencyPackageNode node
+      , none (isDataDependencyPackageNode  . fst3 . vertexToNode0) (reachable depGraph0 v)
+      ]
+
+    -- order the packages in topological order
+    (depGraph0, vertexToNode0, _keyToVertex0) =
+        graphFromEdges $
+            [ (node, pid, pkgRefs)
+            | (isDataDep, (dalf, DecodedDalf{decodedUnitId=unitId, decodedDalfPkg=dalfPackage})) <- fmap (False,) deps <> fmap (True,) dataDeps
+            , let
+                pkg = LF.extPackagePkg (LF.dalfPackagePkg dalfPackage)
+                pid = LF.dalfPackageId dalfPackage
+                pkgRefs = nubOrd [ pid' | LF.PRImport pid' <- allPackageRefs pkg, pid' /= pid ]
+                node
+                  | pid `elem` builtinDeps = MkBuiltinDependencyPackageNode BuiltinDependencyPackageNode {..}
+                  | pid `elem` stablePkgs = MkStableDependencyPackageNode
+                  | isDataDep = MkDataDependencyPackageNode DataDependencyPackageNode {..}
+                  | otherwise = MkDependencyPackageNode DependencyPackageNode {..}
             ]
+
     vertexToNode' v = case vertexToNode v of
         -- We don’t care about outgoing edges.
         (node, key, _keys) -> (node, key)
 
-    dependencyInfo =
-        buildDependencyInfo
-           (map LF.dalfPackagePkg $ MS.elems dependencyPkgs)
-           (LF.initWorld (map (uncurry LF.ExternalPackage) (MS.toList packages)) targetLfVersion)
+data PackageNode
+  = MkDependencyPackageNode DependencyPackageNode
```

Comment:
We used this naming scheme at `$FORMER_EMPLOYER` to avoid partial record selectors